### PR TITLE
Unit tests for opening-saving YAML

### PR DESCRIPTION
**Pull request by KatKatKateryna**
Unit tests added to run PyQt plugin in a headless mode. This required some changes in the main code:
- imports of qgis modules are made optional, as qgis is not required for pytests
- commands requiring qgis types are also made optional (e.g. logging with QgsMessageLog)
- all typing for qgis namespaces are "quoted" in case the code is running from pytests and the type is not imported

Additionally:
- YML workflow is added to run unit tests on every push to a PR branch
- small bug in Provider window (correct widgets' values are passed and read)

Screenshots:
unit_tests_yaml_2

Failed tests:
unit_tests_yaml_failed1
unit_tests_yaml_failed2

### CODE REVIEW

File: tests/pygeoapi-test-config-ogr.yml [hunk 1]
```diff
@@ -0,0 +1,357 @@
+# =================================================================
```
Comment (doublebyte1) [inline tests/pygeoapi-test-config-ogr.yml:1]: Same comment as for the `cite.config.yml`; could we read this file from [here](https://github.com/geopython/pygeoapi/blob/master/tests/pygeoapi-test-config-ogr.yml) instead?

And read `docker.config.yml` from [here](https://github.com/byteroad/ogcapi-simple/blob/master/pygeoapi/docker.config.yml)?

File: tests/test_yaml_save.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import os
+import pytest
+import subprocess
+
+from ..pygeoapi_config_dialog import PygeoapiConfigDialog
+
```
Comment (doublebyte1) [inline tests/test_yaml_save.py:6]: I think we need a third test, to validate that the yaml that we input is the same as the yaml that we output:

Example:

`diff cite.config.yml saved_cite.config.yml ` - The output of this command should be an empty set.

I foresee  some challenges, in order to pass that text (but I am confident you can address them 😁 ):

- The order of the keys may be different in the returned file; nevertheless that should still yield a valid result.
- Some optional fields with default values are being added; that should not happen - we can delegate the task of adding default values to pygeoapi. For example this:
```
    hello-world:
    type: process
    processor:
      name: HelloWorld

```
Is being transformed into this:

```
    hello-world:
        type: collection
        title: ''
        description: ''
        keywords: []
        links: []
        extents:
            spatial:
                bbox: [-180, -90, 180, 90]
                crs: http://www.opengis.net/def/crs/OGC/1.3/CRS84
        providers: []
```

**Review by doublebyte1 — CHANGES_REQUESTED** [review]
Thanks for adding the unit tests, and for including them in the GitHub actions! 🥇 

Could you also add a section about setting up and running the tests (standalone) in the [README file](https://github.com/byteroad/pygeoapi_config/blob/master/README.md)?

I wrote some comments about reading the yaml test files directory from their repositories. Do you think that makes sense, or maybe there is a reason for having them here? 

If we copy them here, I think that we may run into the risk that they become outdated, and we won't even know from which version they belong to. At the same time, reading them from the upstream repositories can also lead to failing tests, and it will be hard to pinpoint if it is because of code changes, or schema changes. I think one other option could also be to tag the config files with a commit (or release) hashtag, and in that way we would know exactly to which version they belong. I would ❤️  to hear your thoughts about this. 

I think with all the code that you introduced, you are in a position to implement the most challenging test of all: comparing the input yaml and the output yaml 😺 

I am aware this may require a refactoring of the code, but from a user's perspective, I think it is super important for the tool to not change anything that was not strictly updated by the user (even when these changes are apparently harmless).

**Comment by doublebyte1** [comment]
> Regarding the files to test against, I would keep them deterministic with the similar set of files to test against. With testing against the versioned config files, what would be that we want to test for exactly?

Hopefully I have answered that question in my other comments. Let me know, if I didn't (:

**Comment by KatKatKateryna** [comment]
Thanks, all clear now! :)

File: tests/yaml_samples/geopython_pygeoapi_@5ed31b4_default.config.yml
Comment (doublebyte1) [inline tests/yaml_samples/geopython_pygeoapi_@5ed31b4_default.config.yml:1]: The temporal extent of lakes collection:

            temporal:
                begin: 2011-11-11
                end: null  # or empty (either means open ended)

Is being converted during the tests to:

            temporal: {}

Maybe because the datetime object is not complete with the timestamp?

In this file we are hitting the problem of a provider that is not supported. As an example, the `dutch_georef_stations` uses the OGR provider. The tests are appending some fields that do not exist in this provider (data:, host: , dbname: user: ') and removing fields that exist for this provider (gdal_ogr_options, source_options, etc).

**Review by doublebyte1 — CHANGES_REQUESTED** [review]
Thanks for adding a script to add the tests locally (and a reference to it in the README): now it is much easier to run the unit tests from the cli. I think that with the sample files you added, we have a variety of configurations that allows us to capture edge cases. 
- Is the convention on the sample yaml file name: `organisation_repository_commit hash`? I think that works pretty well; could you add to the README too? 🙏🏽  
- When I run the tests locally, it correctly generates all the files; however, it does not output if (and which) tests are passed. This is the output: `collected 0 items ========================================= no tests ran in 9.61s ==========================================` . Not sure if this is intended, but shouldn't it compare the input and output files and flag differences as errors?
- There seems to be a problem (at least with some) changes that go through the UI, which are not captured by the tests. The example I noted is the server languages, which have a particular order (that matters). In the case of the unit tests, that order does not change, but if we open the file in the UI and then save it, it does. 
- The non supported providers are still problematic. Ideally, we should not touch those, but if we run the tests, the non supported providers have some keys removed and appended which make them invalid (I have pointed some examples in my comments). If we open these non supported providers in the UI, they are set as read-only, but the validation is still applied.

**Comment by doublebyte1** [comment]
As we discussed, here are the keys that should have free formed strings in the UI (to allow environmental variables):

- server.url
- server.bind.host
- server.bind.port
- On providers.feature.postgresql: data.host, data.port, data.dbname, data.user, data.password

There are examples in the reference file:
https://github.com/byteroad/pygeoapi-config/blob/a015a48e1f7328cff9e070f9a59b764c5bbac2c0/docker.config.yml#L5
https://github.com/byteroad/pygeoapi-config/blob/a015a48e1f7328cff9e070f9a59b764c5bbac2c0/docker.config.yml#L210C1-L215C53
